File: parse_questions.py
import re
import json

def parse_questions(filename):
    with open(filename, 'r', encoding='utf-8') as f:
        content = f.read()

    # Normalize line endings
    content = content.replace('\r\n', '\n')
    
    # We are looking for patterns like "1. What is..." followed by lines until the next number
    pattern = re.compile(r'^(\d+)\.\s+(.*?)\n(.*?)(?=^\d+\.\s+|$)', re.MULTILINE | re.DOTALL)
    
    matches = pattern.findall(content)
    
    questions = []
    
    for match in matches:
        num_str = match[0]
        question_text = match[1].strip()
        answer_text = match[2].strip()
        
        # Clean up answer text by joining lines separated by newline
        answer_text = ' '.join(line.strip() for line in answer_text.split('\n') if line.strip())
        
        # Some cleanup for hyphen separated answers
        # e.g., "Central Processing Unit - the primary component..."
        # If there's a hyphen early on, we can split it to answer and explanation
        split_ans = re.split(r'\s+-\s+', answer_text, maxsplit=1)
        if len(split_ans) == 2:
            answer = split_ans[0].strip()
            explanation = split_ans[1].strip()
            # If the answer part is too long, it might just be a regular sentence with a dash.
            if len(answer) > 80:
                answer = answer_text
                explanation = "A foundational concept in Hardware & Architecture."
        else:
            answer = answer_text
            explanation = "A foundational concept in Hardware & Architecture."

        num = int(num_str)
        if 1 <= num <= 100:
            difficulty = "Easy"
            diff_code = "e"
        elif 101 <= num <= 200:
            difficulty = "Medium"
            diff_code = "m"
        else:
            difficulty = "Difficult"
            diff_code = "d"
            
        q_obj = {
            "id": f"ha-{diff_code}-new-{num}",
            "topic": "Hardware & Architecture",
            "difficulty": difficulty,
            "type": "Identification",
            "question": question_text,
            "answer": answer,
            "explanation": explanation
        }
        
        questions.append(q_obj)
        
    print(f"Parsed {len(questions)} questions.")
    
    # Write to a TS file
    with open('parsed_questions.ts', 'w', encoding='utf-8') as f:
        f.write("import { Question } from \"../services/quizService\";\n\n")
        f.write("export const HARDWARE_NEW_QUESTIONS: Question[] = ")
        # Using string replacement to format nicely
        json_str = json.dumps(questions, indent=2)
        f.write(json_str)
        f.write(";\n")

if __name__ == "__main__":
    parse_questions('questions_raw.txt')

File: parse_questions.js
import fs from 'fs';

function parseQuestions(filename) {
    const lines = fs.readFileSync(filename, 'utf-8').split('\n');
    let questions = [];
    
    let currentNum = null;
    let currentQuestion = '';
    let currentAnswerLines = [];
    
    const saveCurrent = () => {
        if (currentNum !== null && currentQuestion) {
            let answerText = currentAnswerLines.join(' ').trim();
            // remove multiple spaces
            answerText = answerText.replace(/\s+/g, ' ');

            let answer = answerText;
            let explanation = '';

            const dashSplit = answerText.split(/\s+-\s+(.+)/);
            if (dashSplit.length >= 2) {
                const potentialAnswer = dashSplit[0].trim();
                // Check if the answer part is reasonably short
                if (potentialAnswer.length <= 80 && !potentialAnswer.endsWith('.')) {
                    answer = potentialAnswer;
                    explanation = dashSplit[1].trim();
                }
            }
            
            if (!explanation) {
                explanation = "A foundational concept in Hardware & Architecture.";
            }

            // Quick hack for when it doesn't split well
            if (answer.length > 100) {
               // If answer is too long, we can shorten it and put the rest in explanation
               const periodIndex = answer.indexOf('.');
               if (periodIndex !== -1 && periodIndex < 100) {
                   explanation = answer.substring(periodIndex + 1).trim() + " " + explanation;
                   answer = answer.substring(0, periodIndex + 1);
               }
            }
            
            let difficulty = "Difficult";
            let diffCode = "d";
            
            if (currentNum >= 1 && currentNum <= 100) {
                difficulty = "Easy";
                diffCode = "e";
            } else if (currentNum >= 101 && currentNum <= 200) {
                difficulty = "Medium";
                diffCode = "m";
            }
            
            questions.push({
                id: `ha-${diffCode}-new-${currentNum}`,
                topic: "Hardware & Architecture",
                difficulty: difficulty,
                type: "Identification",
                question: currentQuestion,
                answer: answer,
                explanation: explanation
            });
        }
    };
    
    for (let i = 0; i < lines.length; i++) {
        let line = lines[i].trim();
        if (!line) continue;
        
        const questionMatch = line.match(/^(\d+)\.\s+(.*)/);
        if (questionMatch) {
            saveCurrent();
            currentNum = parseInt(questionMatch[1], 10);
            currentQuestion = questionMatch[2];
            currentAnswerLines = [];
        } else if (currentNum !== null) {
            // Check if it's not a heading like "EASY LEVEL (Questions 1-100)"
            if (line.match(/^[A-Z]+ LEVEL/)) continue;
            if (line.match(/^[A-Za-z]+ \(\d+-\d+\)/)) continue;
            if (line.match(/^Basic/i) && line.indexOf('1-10') > -1) continue;
            currentAnswerLines.push(line);
        }
    }
    
    // Save last question
    saveCurrent();
    
    console.log(`Parsed ${questions.length} questions.`);
    
    const output = `import { Question } from "../services/quizService";\n\nexport const HARDWARE_NEW_QUESTIONS: Question[] = ${JSON.stringify(questions, null, 2)};\n`;
    fs.writeFileSync('parsed_questions.ts', output, 'utf-8');
}

parseQuestions('questions_raw.txt');
